### PR TITLE
[RFC] Add new epmd application (written in Erlang)

### PR DESCRIPTION
**Pull request by lemenkov**
This patch adds entirely new EPMD, written in Erlang. My plan is to replace an old C-based one  with this one entirely (with some additional work).

What doesn't work right now:

* IPv6 probably doesn't work.
* -daemon option
* Additional 'packet_timeout' debug option.
* No tests are yet available (I plan to reuse ones from old epmd w/o any changes).

I'm not sure if it's ready for merge but I submitted it anyway in hope to get some feedback from others.

cc @doubleyou @msantos @stolen

### CODE REVIEW

**Comment by essen** [comment]
Very nice!

Just gave a quick look and have an immediate comment so far: the module name `tcp_listener` will clash with some existing code, like RabbitMQ. What about `epmd_tcp_listener`?

**Comment by michaelklishin** [comment]
I don't know if epmd nodes will ever run code other than epmd itself but RabbitMQ indeed has a module named `tcp_listener`.

**Comment by doubleyou** [comment]
Looks great!

One thing that came to my mind: how can we make sure releases are compatible with this `epmd` implementation? Maybe escriptize the whole application and ship it along with ERTS?

**Comment by okeuday** [comment]
@lemenkov Doesn't `-heart` usage with the Erlang VM require that epmd is not using the Erlang VM?

**Comment by lemenkov** [comment]
Ok, IPv6 seems working

    sh bin/erlpmd.sh -address 192.168.1.201,2002:XXXX:XXXX:XXXX:426c:8fff:fe36:30a

Unfortunately I can't test it with a stock epmd from erlang-17.4 (looks like it doesn't support IPv6).

**Comment by lemenkov** [comment]
See also issues #602, #708

**Comment by OTP-Maintainer** [comment]
Fetching pull requests from github failed:

    Host key verification failed.
    fatal: Could not read from remote repository.
    Please make sure you have the correct access rights
    and the repository exists.



----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Fetching pull requests from github failed:

    Permission denied (publickey).
    fatal: Could not read from remote repository.
    Please make sure you have the correct access rights
    and the repository exists.



----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Fetching pull requests from github failed:

    Permission denied (publickey).
    fatal: Could not read from remote repository.
    Please make sure you have the correct access rights
    and the repository exists.



----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Fetching pull requests from github failed:

    Permission denied (publickey).
    fatal: Could not read from remote repository.
    Please make sure you have the correct access rights
    and the repository exists.



----------------

I am a script, I am not human 

----------------

**Comment by proxyles** [comment]
Sorry for the spam! I have trolls here

**Comment by OTP-Maintainer** [comment]
Fetching pull requests from github failed:

    Permission denied (publickey).
    fatal: Could not read from remote repository.
    Please make sure you have the correct access rights
    and the repository exists.



----------------

I am a script, I am not human 

----------------

**Comment by OTP-Maintainer** [comment]
Fetching pull requests from github failed:

    Permission denied (publickey).
    fatal: Could not read from remote repository.
    Please make sure you have the correct access rights
    and the repository exists.



----------------

I am a script, I am not human 

----------------

**Comment by proxyles** [comment]
Ok. Ill dissable and dig abit more. Again, sorry for the spam

**Comment by OTP-Maintainer** [comment]
Patch has passed first testings and has been assigned to be reviewed


----------------

I am a script, I am not human 

----------------

**Comment by proxyles** [comment]
Ok then!

**Comment by stolen** [comment]
Great work, Peter!
The code seems to be written for some testing purposes — all these optional delays, low-level accepting, etc.
Maybe it would be more useful to separate protocol implementation from behaviour tweaks. I'm thinking of something like `gen_epmd` with 
``` erlang
gen_epmd:start_link(HostName, Port, CallbackMod, CallbackOpts, Options) -> {ok, pid()}.
gen_epmd:accept(Socket) -> async.

CallbackMod:init(HostName, Port, CallbackOpts) -> {ok, State}.
CallbackMod:handle_connection(Socket, State) -> {accept, NewState} | {reject, NewState} | {hold, NewState}.
CallbackMod:handle_alive(NodeName, PortNo, #{} = AllOtherStuff, State) -> {reply, {ok, Creation}, NewState}.
```
and so on.

A simple (default) behaviour may be shipped with OTP while allowing users to easily implement their own EPMDs with all the tweaks they need.

**Comment by lemenkov** [comment]
@stolen yep, this makes sense. I'll look into it.

**Comment by neeraj9** [comment]
Awesome! I was in the middle of integrating in my Erlang port to rumprun unikernels, while someone pointed me here. It was very easy for me to integrate this pure Erlang rewrite. Great work!

**Comment by lemenkov** [comment]
@neeraj9 Awesome! Glad to hear that it wasn't entirely useless :)
Please, give me a few more days to finish everything planned. I believe we'll have something ready for integration to the end of September (maybe earlier).

**Comment by neeraj9** [comment]
@lemenkov I have integrated your patch already, since it's usefulness outweighs the side-effects and then things can always be fixed and reloaded ;)
Would be great to see this going mainstream.

File: lib/epmd/src/erlpmd.erl [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	error_logger:warning_msg("ErlPMD: strange call: ~p from: ~p.~n", [Request, From]),
+	{reply, ok, State}.
+
+handle_cast({{msg,From},<<?EPMD_ALIVE2_REQ, PortNo:16, NodeType:8, Proto:8, HiVer:16, LoVer:16, NLen:16, Rest/binary>>, Fd, Ip, Port}, State = #argv{delay_write = DelayWrite}) ->
+	<<NodeName:NLen/binary, _ELen:16, Extra/binary>> = Rest,
+	Creation = random:uniform(3),
+	error_logger:info_msg(
+		"ErlPMD: alive request from ~s:~b PortNo: ~b, NodeType: ~b, Proto: ~b, HiVer: ~b, LoVer: ~b, NodeName: '~s', Extra: ~p, Creation: ~b.~n",
+		[inet_parse:ntoa(Ip), Port, PortNo, NodeType, Proto, HiVer, LoVer, NodeName, Extra, Creation]),
+	case ets:lookup(erlpmd, NodeName) of
```
Comment (juise) [inline lib/epmd/src/erlpmd.erl]: May be better will be:
```
case ets:insert_new(erlpmd, {NodeName, {PortNo, NodeType, Proto, HiVer, LoVer, Extra, Creation, Fd}}) of
  true ->
    reply(From, {msg, <<?EPMD_ALIVE2_RESP, 0:8, Creation:16>>, Ip, Port}, DelayWrite);
  false ->
    error_logger:error_msg("ErlPMD: ~s 'name' is already registered.~n", [NodeName]),
    reply(From, {msg, <<?EPMD_ALIVE2_RESP, 1:8, 99:16>>, Ip, Port}, DelayWrite)
end,
```

**Comment by peerst** [comment]
This is a nice example of lack of coordination for these things.

I have a pure Erlang epmd running in production in embedded systems for 1.5years now but was too humble to just try a pull request on OTP without talking if they even want it.  My own fault, now we have two of those beasts (apart from the probably several others built in tutorials).

I'd be willing to submit my code too, it has seen less activity recently since it just works and lacks IPv6 which is trivial to add.

My code is at:

https://bitbucket.org/peerst/epmd

**Comment by peerst** [comment]
Regarding the possible name clashes that were mentioned by Michael Klishin with the comment that a Erlang epmd would never run aside normal code.  This is a assumption that doesn't hold:  I wrote my epmd to be started by some spceial code in net_kernel in a normal Erlang node.  For running on normal Unix like operating systems this doesn't make sense but the whole reason I wrote this epmd is to run in a small embedded realtime operating system (RTEMS) where there can only be one Erlang VM per device anyway.

**Comment by saleyn** [comment]
I am curious what was the need to rewrite an epmd in Erlang vs living with the legacy one?

**Comment by peerst** [comment]
Not sure if you mean my implementation or this one.  

In my case Erlang is running on RTEMS a embedded realtime executive which only has threads and no processes.  Everything is linked together there.

A similar setup was using the legacy epmd on Vxworks with some hackery to make it run alongside Erlang in a thread.  However the Vxworks port was unused when I did my port to RTEMS and I didn't want to mess with the C epmd implementation.

Also I talked to Patrik Nyblom who did the work on epmd for Vwworks and he said that he would probably  implement epmd in Erlang for a similar setup.

I can run epmd in the same node that also runs the application code. There is no possibility to start epmd twice this way since there can be only one Erlang Node per RTEMS Node.

**Comment by okeuday** [comment]
@saleyn It appears that epmd is a single point of failure for distributed Erlang, so it should be beneficial for the fault-tolerance of the Erlang VM if epmd is moved completely into Erlang source code.  At the very least, https://github.com/erlang/otp/pull/1022 will make it simpler to avoid the use of the current epmd.

**Comment by KennethL** [comment]
I agree that it is a good idea to have Epmd implemented in Erlang. But I
don't understand when the current epmd implementation has been a point of
failure?

The benefit from having Epmd in Erlang would be a more productive and
robust way to implement new features, especially during prototyping.

I actually think of many use cases of Erlang distribution where Epmd is not
needed at all ( knowing exactly what epmd is used for).

/Kenneth,  Erlang/OTP Ericsson
Den 22 apr 2016 19:48 skrev "Michael Truog" <notifications@github.com>:

**Comment by okeuday** [comment]
@KennethL Epmd in its current form should be a single point of failure for distributed Erlang due to the possibility that the separate OS process executing epmd is terminated or suspended, which should cause an interruption to what epmd provides (epmd is mentioned as a single point of failure at http://stackoverflow.com/a/19515024 ).  If epmd is moved into Erlang, it should have better liveness guarantees due to the Erlang VM with less development effort than might be required to make the current epmd survive different failure scenarios.

**Comment by peerst** [comment]
The single point of failure is theoretical though.  Yes its a single process that when killed forgets about erlang nodes.  But why would it be killed?  If you or your system randomly kills processes you have more points of failure pretty soon.

Its quite stable for a long time and won't crash on you easily by itself.

Being more flexible when extending I totally get.

Also I obviously get the embedded systems use case where porting epmd is a hassle for very small systems because thats why I implemented my version of it in Erlang.

**Comment by okeuday** [comment]
@peerst While you may regard the possibility of an epmd failure as theoretical and only due to external influence, there also is the consideration of potentially adding bugs to the epmd source code as epmd may change in the future.  Handling the transient errors that developers may create will be handled better with epmd as Erlang source code.

**Comment by peerst** [comment]
Everyone including myself agreed already that for extension its better in Erlang.

**Comment by OTP-Maintainer** [comment]
The summary line of the commit message is too long and/or ends with a "."
Make sure the whole message follows the guidelines here: https://github.com/erlang/otp/wiki/Writing-good-commit-messages.

Bad message: Use erlang:apply/3 instead of calling optional library directly


----------------

I am a script, I am not human 

----------------

**Comment by psyeugenic** [comment]
We should discuss a way forward ..

My take on this is that we should *not* include an EPMD application written in Erlang in the main OTP repository. Instead it should be a standalone repository and in this case I think it should reside in the Erlang organization here at GitHub. We definitely do want to move away from the C implementation to an Erlang implementation. 

I've looked at both @lemenkov code here and @peerst EPMD application. Both has its pros and cons.
Regarding @lemenkovs PR: Pretty feature complete except for the things you stated. However, I take issue with the coding style but that is perhaps a matter of taste. Also, the optional call to `sd_notify`, I don't really like it. I understand why you do it that way, it's optional but there is no mention of where I can find that module.

I looked a lot more on @peerst EPMD because it looked nicer. However it lacks certain things that @lemenkov has implemented here. I took @peerst code, added test cases and fix most of the failing ones. It has a couple of issues still. You can view it here: https://github.com/psyeugenic/epmd

So now we have three repositories ..

**Comment by peerst** [comment]
Excellent, I'll switch using this new epmd for my embedded use case ASAP.  Will point from my bitbucket epmd repo to your and only PR new changes to https://github.com/psyeugenic/epmd

So now we have two repositories again ;-)

**Comment by lemenkov** [comment]
Ok, I'm going to close this PR since upstream version will be available.

**Pull request closed without merging**